src/org/labkey/targetedms/model/GuideSetStats.java

```diff
@@ -0,0 +1,164 @@
+/*
+ * Copyright (c) 2019 LabKey Corporation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.labkey.targetedms.model;
+
+import org.labkey.api.visualization.Stats;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class GuideSetStats
+{
+    private final GuideSetKey _key;
+    private final GuideSet _guideSet;
+
+    /** Rows that define the normal range for this guide set */
+    private final List<RawMetricDataSet> _trainingRows = new ArrayList<>();
+    /** Rows that reference the training values */
+    private final List<RawMetricDataSet> _referenceRows = new ArrayList<>();
+
+    public GuideSetKey getKey()
+    {
+        return _key;
+    }
+
+    private double _standardDeviation;
+    private double _average;
+
+    private double _movingRangeAverage;
+
+    private boolean _locked = false;
+
+    public GuideSetStats(GuideSetKey key, GuideSet guideSet)
+    {
+        _key = key;
+        _guideSet = guideSet;
+    }
+
+    public double getStandardDeviation()
+    {
+        assertLocked();
+        return _standardDeviation;
+    }
+
+    private void assertLocked()
+    {
+        if (!_locked)
+        {
+            throw new IllegalStateException("Stats have not yet been calculated");
+        }
+    }
+
+    public double getAverage()
+    {
+        assertLocked();
+        return _average;
+    }
+
+    public double getMovingRangeAverage()
+    {
+        assertLocked();
+        return _movingRangeAverage;
+    }
+
+    public void addRow(RawMetricDataSet row)
+    {
+        if (_locked)
+        {
+            throw new IllegalStateException("Stats have already been locked");
+        }
+        if (!row.isIgnoreInQC() &&
+                _guideSet.getTrainingStart().compareTo(row.getAcquiredTime()) <= 0 &&
+                (_guideSet.getTrainingEnd() == null || _guideSet.getTrainingEnd().compareTo(row.getAcquiredTime()) >= 0))
```

Comment:
Could row.getAcquitedTime ever be null here?

Reply:
No, it's now filtered out in the query. A test had a case where one sample file has a null acquired time, and without any idea of where to put it in the time sequence, it's not useful for QC purposes